Assets/Scripts/Core/Puzzles/PuzzleController.cs

```diff
@@ -0,0 +1,188 @@
+					puzzleBlockScript.OnPuzzleBlockClicked += OnPlayerMoveBlockInput;
+					puzzleBlockScript.OnPuzzleBlockFinished += OnPuzzleBlockFinished;
+					puzzleBlockScript.Init(puzzleSlices[x, y], new Vector2Int(x, y));
+
+					puzzleBlocks_[x, y] = puzzleBlockScript;
+
+					// This removes the bottom right block of the puzzle
+					if(y == 0 && x == blocksPerLine_ - 1) {
+						emptyPuzzleBlock_ = puzzleBlockScript;
+						puzzleBlock.SetActive(false);
+					}
+				}
+			}
+			input_ = new Queue<PuzzleBlock>();
+			GetComponent<Image>().enabled = true;
+			StartShuffle();
+		}
+
+		public void DestroyPuzzle() {
+			foreach(Transform child in transform) {
+				Destroy(child.gameObject);
+			}
+			GetComponent<Image>().enabled = false;
+		}
+
+		public void OnPlayerMoveBlockInput(object sender, PuzzleBlock block) {
+			if(currentState_ != PuzzleState.Active) {
+				return;
+			}
+			input_.Enqueue(block);
+			MakeNextMove();
+		}
+
+		public void OnPuzzleBlockFinished(object sender, System.EventArgs e) {
+			blockIsMoving_ = false;
+
+			if(currentState_ == PuzzleState.Active) {
+				if(IsPuzzleSolved()) {
+					currentState_ = PuzzleState.Solved;
+					emptyPuzzleBlock_.gameObject.SetActive(true);
+					puzzleVictory_.gameObject.SetActive(true);
+				} else {
+					MakeNextMove();
+				}
+			}
+			
+
+			if(shuffleCountRemaining_ > 0) {
+				MakeNextShuffleMove();
+			} else {
+				currentState_ = PuzzleState.Active;
+			}
+		}
+
+		private void MoveBlock(PuzzleBlock block, float duration) {
+			// Check if block is adjacent to empty block
+			if((block.coord - emptyPuzzleBlock_.coord).sqrMagnitude == 1) {
+				// Swap blocks
+				puzzleBlocks_[block.coord.x , block.coord.y] = emptyPuzzleBlock_;
+				puzzleBlocks_[emptyPuzzleBlock_.coord.x, emptyPuzzleBlock_.coord.y] = block;
+
+				Vector2Int emptyCoord = emptyPuzzleBlock_.coord;
+				emptyPuzzleBlock_.coord = block.coord;
+				block.coord = emptyCoord;
+
+				Vector3 targetPosition = emptyPuzzleBlock_.transform.position;
+				emptyPuzzleBlock_.transform.position = block.transform.position;
+				block.MoveToPostion(targetPosition, duration);
+				blockIsMoving_ = true;
+			}
+		}
+
+		private void MakeNextMove() {
+			while(input_.Count > 0 && !blockIsMoving_) {
+				MoveBlock(input_.Dequeue(), defaultMoveTime_);
+			}
+		}
+
+		private void StartShuffle() {
+			shuffleCountRemaining_ = shuffleCount_;
+			currentState_ = PuzzleState.Inactive;
+			MakeNextShuffleMove();
+		}
+
+		private void MakeNextShuffleMove() {
+			Vector2Int[] directions = new Vector2Int[] {
+				new Vector2Int(0, 1),	// Up
+				new Vector2Int(1, 0),	// Right
+				new Vector2Int(0, -1),	// Down
+				new Vector2Int(-1, 0)	// Left
+			};
+
+			int randomIndex = Random.Range(0, directions.Length);
+
+			for(int i = 0; i < directions.Length; i++) {
+				Vector2Int randomDirection = directions[(randomIndex + i) % directions.Length];
+				if(randomDirection != prevShuffle_ * -1) {
```

Comment:
Coding style violation, missing space between `-` and `1`.

Reply:
Same here.

Reply:
Yep, missed the operator, whoops!